src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/TypeNameBuilder.cs

```diff
@@ -0,0 +1,370 @@
+        {
+            Append('*');
+        }
+
+        public void AddByRef()
+        {
+            Append('&');
+        }
+
+        public void AddSzArray()
+        {
+            Append("[]");
+        }
+
+        public void AddArray(int rank)
+        {
+            if (rank <= 0)
+                throw new ArgumentOutOfRangeException();
+
+            if (rank == 1)
+            {
+                Append("[*]");
+            }
+            else if (rank > 64)
+            {
+                // Only taken in an error path, runtime will not load arrays of more than 32 dimensions
+                _str.Append('[').Append(rank).Append(']');
+            }
+            else
+            {
+                Append('[');
+                for (int i = 1; i < rank; i++)
+                    Append(',');
+                Append(']');
+            }
+        }
+
+        public void AddAssemblySpec(string assemblySpec)
+        {
+            if (assemblySpec != null && !assemblySpec.Equals(""))
+            {
+                Append(", ");
+
+                if (_instNesting > 0)
+                {
+                    EscapeEmbeddedAssemblyName(assemblySpec);
+                }
+                else
+                {
+                    EscapeAssemblyName(assemblySpec);
+                }
+
+                _hasAssemblySpec = true;
+            }
+        }
+
+        public override string ToString()
+        {
+            Debug.Assert(_instNesting == 0);
+
+            return _str.ToString();
+        }
+
+        private static bool ContainsReservedChar(string name)
+        {
+            foreach (char c in name)
+            {
+                if (c == '\0')
+                    break;
+                if (IsTypeNameReservedChar(c))
+                    return true;
+            }
+            return false;
+        }
+
+        private static bool IsTypeNameReservedChar(char ch)
+        {
+            switch (ch)
+            {
+                case ',':
+                case '[':
+                case ']':
+                case '&':
+                case '*':
+                case '+':
+                case '\\':
+                    return true;
+
+                default:
+                    return false;
+            }
+        }
+
+        private void EscapeName(string name)
+        {
+            if (ContainsReservedChar(name))
+            {
+                foreach (char c in name)
+                {
+                    if (c == '\0')
+                        break;
+                    if (IsTypeNameReservedChar(c))
+                        _str.Append('\\');
+                    _str.Append(c);
+                }
+            }
+            else
+                Append(name);
+        }
+
+        private void EscapeAssemblyName(string name)
+        {
+            Append(name);
+        }
+
+        private void EscapeEmbeddedAssemblyName(string name)
+        {
+            bool containsReservedChar = false;
```

Comment:
```suggestion
            bool constainsReservedChar = name.Contains(']');
```

You could optimize it further with IndexOf